### ISSUE TITLE
[Feature request] Getting global transform of a bone should be easier.

### ISSUE THREAD

**Issue by slapin** [urho3d/Urho3D#1840]
Urho bones have "local global" namespace, so they are really hard to work with when something complex concerned.
Some method on Bone is needed to get actual global transform or transform in Skeleton node space.
As I have to write this code every now and then and see people asking about this, I think it is first serious
beginner's obstacle with Urho, which really confuses people. Also this one is easy enough to implement.
Also reversed method (converting global space/skeleton space to bone space) should be very nice to have too.

**Comment by cadaver** [comment]
I'm not sure I follow. The bone nodes are like any scene nodes in the scene, in that you can operate in either local or global space, and this should be similar to any engine operating on a scene graph.

It's true that working automatically in "skeleton space" is missing. To clarify, do you mean the space that's calculated with (pseudocode) AnimatedModel.WorldTransform.Inverse * Bone.WorldTransform?

**Comment by slapin** [comment]
> To clarify, do you mean the space that's calculated with (pseudocode)
AnimatedModel.WorldTransform.Inverse * Bone.WorldTransform?
Well, more or less yes.

I think I need to explain farther what concerns me, as there looks to be
misunderstanding.
I never had experience with gamedev or engines, or much 3D graphics, as my
field of expertise is embedded systems,
system programming, and Linux kernel. So I might lack terminology knowledge
here, but I will try to use all my experience
to explain what I want. Also don't think I really know what I'm doing, my
lack of game dev experience might lead me
into wrong direction but I really believe I'm on right track at the moment.

Well, the main problem I have now is the following: (I don't have
experience with other engines to consider anything
being similar).

Imagine having Skeleton with ~170 bones (20 * 4 = hand and feet + simple
face rig + rest of body) of about this topology:

https://thomasmakehuman.files.wordpress.com/2015/08/140-imported-rigify.png

And you need to implement the folowing tasks for it to reduce amount of
animation work and increase
graphics quality:

1. Implement ragdoll, which is capable to transition from pose to ragdoll
and back into pose.
So the ragdoll state should be reversible in fluent manner. This includes
partial ragdoll, too.

2. Implement corrective IK for legs to not penetrate ground and for arms
motion to the goal and aiming.
(2 character holding arms, aiming according to cursor).

Both cases assume complex animations with "not deterministic" state
(animation is layered and depends on many factors).

For both scenarios you need to have values calculated which depend on limb
lengths and lengths of limb parts.
I consider using skeleton with the same topology but different proportions
and sizes.
For ragdoll:
The state in which skeleton currently is is unpredictable. We need to
convert it to ragdoll without changing visual
appearance immediately, so to setup standard joints we need to setup
constraints, i.e. to
avoid head explosion, we need to work in skeleton node space to have right
joint limits (as body spaces are weird here),
so we need to implement calculations which convert from skeleton space to
each particular body space.

Also the same kind of calculation applies when we need to setup collision
shapes for the bodies.

Things become a bit more complex if we want to setup this once and then
periodically enable/disable it,
still blending with animation. The most "hot" data needed are:
1. Simple conversion from skeleton space to each bone space. This is
complicated by need to look-up bone in Skeleton,
then get one from node tree via GetChild which doesn't look clear enough.
2. Easy method to find dimentions for bones not relying on exporter (which
never worked for me) to provide relevant data.

For IK:
To solve single-joint IK we can assume it is flat and solve in plane (2D)
which is directly solvable. To have this working we need sizes for each
our complex bone (consisting of 2 bones in our case). The sizes can be
obtained by offset difference of respective
differences in transform origins. This requires conversion of this set of
transforms to skeleton space to provide calculations needed.
In addition to support independent goals for IK the goal have to be
transformed to skeleton space, but that can be handled directly.

The main issue is that addressing both of these problems (especially at the
same time, like in my case) produce extremely hairy code
(which is also not really fast). As we all can do everything in OpenGL
directly, but want engine to save us the trouble,
and make complex things simpler, I think something should be done about
this so that bone manipulations get more intuitive
and less complex.

**Comment by JSandusky** [comment]
I don't see the problem. 

Though I'll admit that I can't understand what exactly you're trying to say. You're practically enumerating features for the first 50%, the top stuff is easily doable.

I use Urho3D's skeleton functionality for solving IK and remapping animation between different skeletal morphologies without issue. It all works, a bone's global and local transform is easily accessed. That's the very title of your post, but it's so easy to get without any matrix ops.

Instead of long-winded cases, what is it that you actually need to get access to? What the title of this post says is easily accessed.

If this is your first foray into IK, then yes, you have to "eat" it, for a whole lot of things.

**Comment by cadaver** [comment]
Another thing to note is, that when you apply world space simulation / IK results back to a bone hierarchy using SetWorldPosition() etc. you need to do this in parent -> child order, because SetWorldPosition() just modifies the local position too; it first calculates what it should be based on the current parent transform. The Physics subsystem does this automatically when it's applying node transforms back from the rigidbodies, which is demonstrated in the ragdoll sample.

**Comment by slapin** [comment]
Well, I'm very sorry about allowing my mind to fly around when I write
reports. I'm not really experienced with other engines, so
I don't know patterns here so I probably on wrong path, but the right path
is too hard to find. Anyway, limiting this to the point.

Well, to many tasks working with seketons I find the following difficulties:

As designer designed the character and told the set of bones which need
special attention (bone names).
Also one have to use (possibly large) number of bones for ragdoll. All this
requires conversion of bone name to bone node.
So we have a set of bone names as input (which might be large).

1.  The frequent problem is conversion from Skeleton stuff to node tree
stuff. The bone handling is very verbose because of that -
we pick bone from Skeleton and then looking-up node in hierarchy all the
time. Why not accelerate this operation?

2. Another issue I have (and hate most) is with constant need of conversion
from global space to local space to skeleton space.
Imagine setting-up ragdoll. It is much easier to have data in sekeleton
space - rotations for constraints and collision shapes.
Collision shape rotations are nice to be local unless something complicated
is designed, but for a few dozen bones it would be
 nice to have these rotations
in skeleton space. Yes, you can do all this in your own code, but I think
engine could be great help here. My suggestion here would be to
use now underpowered Skeleton class.
3. It would be nice to have reference from Bone to node and from Node to
Bone quickly, that would help cleaning the code a lot.

4. Having access to bone rest position in local, global and skeleton space
would help me a great deal too.

This is probably as close to what I can currently grasp as possible.
Sorry for vague descriptions - I was emotionally unstable then. Now I
basically managed to have some things
working, like behavior trees running and I'm much more happy user of Urho
now. Now I want my ragdolls to scale
(so setting-up new one would not turn into grotesque nightmare) and bones
would help a lot (imagine, the most huge
and unreadable code in my game is ragdoll setup!). I evn ca't dream about
IK setup now as nothing I tried really working,
so I hope for 1 things at time. I just wonder how Unity do it - I tried
their wizard lately and it got my ragdoll right with a few clicks,
with Urho it took me several months to do it (mostly due to this
entertaining transform conversion). I just really hope for some
improvement on this front.

**Comment by JSandusky** [comment]
Thanks for more clarification.

*Item #1:* I see the problem, it would be handy if AnimatedModel components kept a WeakPtr<Node> to their child nodes that were produced from skeleton bones. I believe that would also address Item #3 as it'd cover Bone->Node mapping (with a utility function for the Node->Bone translation as that's not something that can be naturally represented in the scene-graph at present).

*Item #2:* You mean "model-space" right, or do you actually mean skeleton space? (that would be root bone is origin).

Bones in the scene aren't particularly special (beyond being classified as "temporaries"), but that could be addressed with some utility functions, something along the lines of:

    Matrix3x4 Node::GetRelativeTransform(Node* otherNode) const;
    Matrix3x4 Node::GetRelativeTransform(const Matrix3x4& otherTransform) const;

To reduce the error risk and verbosity involved in repetitive nodeA->GetWorldTransform().Inverse() * nodeB->GetWorldTransform(). That would also deal with other space<->otherSpace conversions aside from just bones (such as "socket" snapping).

*Item #4:* That's still ambiguous. These spaces are potentially very different (actual object instance in world vs. the "clean-room" definition, etc).

**Comment by slapin** [comment]
As "skeleton space" I mean model space.

**Comment by cadaver** [comment]
If you get the Skeleton object from the AnimatedModel instance (as opposed to the Model), there already are WeakPtr node pointers inside the Bone objects, so you don't have to lookup from scene separately. Should also be accessible from script.

**Comment by slapin** [comment]
The problem is that these pointers do not represent hierarchy so there's no
vay to avoid lookups. Also, it doesn't workin reverse :(

**Comment by slapin** [comment]
Well, I almost forgot - setting-up constraint on bones is nightmarish task currently.
That is why. As this involves:

1. Loop over all bones involved in ragdoll (i.e. you have 300 and ragdoll have 20).
2. For each bone convert model space to bone space (no single-step method).
3. Setup RigidBodies using either local or model space (depending on bone or model complexity,
or taste).
4. Setup constraints converting data from model space to bone space. It is really hard to use bone local
space for constraints with real world models.

If this is not scary try to convert ragdoll example to using model space - you will see how hairy it becomes
(if it is not hairy enough now). This look like API could really help here.
I would really-really be happy for some helper here. Or wizard in editor. Or whatever.
I tried to make my own but I failed to grasp editor and GUI system.
As we already have some complicated components, I think some ragdoll helper or just some bone constraint helper should not be a problem and won't make API too hacky.
